### PR TITLE
APP-2072 Migrate PRIME to Svelte 4.

### PR DESCRIPTION
**Pull request by ethanlook**
@mcous thanks for the inspiration for PR descriptions :)

## Overview

This migrates PRIME to Svelte 4. I wanted to do this migration before changing the package structure.

## Change log

- `npx svelte-migrate@latest svelte-4`
- Fix package version mismatches
- Import styles directly instead of `addStyles` helper
- Fix button and input `attachInternals`
- Fix code editor injected styles
- Change slider `range` prop to only accept a string - mixed types doesn't play well with web components
- Use Svelte `createEventDispatcher` directly - `get_current_component` no longer works
- Fix problematic tests

## Review requests

Pull down the branch and make sure:
- `pnpm run test` passes
- Elements look/behave correctly with `pnpm run storybook` and `pnpm start`

Questions:
- Is the minor version bump to 0.3.0 appropriate? I imagine we'll bump to 1.0.0 with the package restructure since that will be breaking.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -76,22 +76,22 @@
...
     "monaco-editor": "^0.37.1",
     "postcss": "^8.4.23",
     "prettier": "^2.8.8",
-    "prettier-plugin-svelte": "^2.10.0",
+    "prettier-plugin-svelte": "^2.10.1",
     "prismjs": "^1.29.0",
     "react": "^18.2.0",
     "remark-gfm": "^3.0.1",
     "storybook": "^7.0.7",
-    "svelte": "^3.58.0",
```
Comment (micheal-parks) [inline package.json:90]: Can we upgrade to latest? There are some patch upgrades out 🕺

File: package.json [hunk 2]
```diff
@@ -76,22 +76,22 @@
...
-    "svelte-check": "^3.2.0",
+    "svelte": "^4.0.0",
+    "svelte-check": "^3.4.3",
     "svelte-preprocess": "^5.0.3",
     "tailwindcss": "^3.3.1",
-    "typescript": "^5.0.4",
```
Comment (micheal-parks) [inline package.json:94]: y go back?

File: src/elements/tooltip.svelte [hunk 3]
```diff
@@ -24,7 +23,7 @@ $: {
 let x = 0;
 let y = 0;
 
-export const recalculateStyle = async () => {
+export let recalculateStyle = async () => {
```
Comment (micheal-parks) [inline src/elements/tooltip.svelte:26]: Why the change to `let`?

Comment (ethanlook) [reply src/elements/tooltip.svelte:26]: Sadly, using `const` produces a runtime error on the `create_custom_element` call:

```
TypeError: Cannot redefine property: recalculateStyle
```

**Comment by mcous** [comment]
> thanks for the inspiration for PR descriptions :)

The template is here for the taking, if we want it!
<https://github.com/Opentrons/opentrons/blob/edge/.github/PULL_REQUEST_TEMPLATE.md>

**Comment by DTCurrie** [comment]
> > thanks for the inspiration for PR descriptions :)
> 
> The template is here for the taking, if we want it! https://github.com/Opentrons/opentrons/blob/edge/.github/PULL_REQUEST_TEMPLATE.md

If there is no opposition to it I don't mind setting it as our template.

File: package.json (hunk 2 above)
Comment (ethanlook) [reply package.json:94]: I believe this was changed by the svelte upgrade script. I upgraded to the latest.

File: src/elements/tooltip.svelte (hunk 3 above)
Comment (micheal-parks) [reply src/elements/tooltip.svelte:26]: ahhh interesting

File: .eslintrc.cjs [hunk 4]
```diff
@@ -5,6 +5,7 @@ module.exports = {
   env: {
     browser: true,
   },
+  parser: '@typescript-eslint/parser',
```
Comment (mcous) [inline .eslintrc.cjs:8]: [Nit] I think this might be set by `@viamrobotics/eslint-config`, and may not be needed. However, it is not harmful to include

File: src/elements/badge.svelte [hunk 5]
```diff
@@ -25,3 +22,7 @@ addStyles();
 >
   {label}
 </small>
+
+<style>
+@import './../../prime.css';
```
Comment (mcous) [inline src/elements/badge.svelte:27]: [nit] does `./..` === `../`?

```suggestion
@import '../../prime.css';
```

File: src/elements/badge.svelte [hunk 6]
```diff
@@ -25,3 +22,7 @@ addStyles();
 >
   {label}
 </small>
+
+<style>
```
Comment (mcous) [inline src/elements/badge.svelte:26]: Is this running through PostCSS? I thought [this was necessary](https://tailwindcss.com/docs/guides/sveltekit):

```suggestion
<style lang="postcss">
```

**Comment by ethanlook** [comment]
Bundle size is down with Svelte 4!

Before:
```
dist/prime.css   25.84 kB │ gzip:  5.49 kB
dist/prime.js   329.75 kB │ gzip: 72.66 kB
dist/prime.umd.cjs  210.17 kB │ gzip: 59.10 kB
```

After:
```
dist/prime.css   25.84 kB │ gzip:  5.49 kB
dist/prime.js   318.95 kB │ gzip: 71.46 kB
dist/prime.umd.cjs  190.12 kB │ gzip: 56.52 kB
```

**Review by mcous — APPROVED** [review]
Tests and playground LGTM!

> Is the minor version bump to 0.3.0 appropriate? I imagine we'll bump to 1.0.0 with the package restructure since that will be breaking.

Works for me. Since we're pre 1.0, `0.3.0` is _not_ semver compatible with `^0.2.0`, which should cover our bases

**Review by micheal-parks — APPROVED** [review]
🎉 glad we arrived at this solution

File: src/main.ts [hunk 7]
```diff
@@ -5,44 +5,91 @@ import '../prime.css';
...
+ * https://github.com/sveltejs/svelte/blob/master/packages/svelte/src/runtime/internal/public.d.ts#L72-L73
+ */
+type ComponentWithElement = SvelteComponent & {
+  element: typeof HTMLElement;
+};
+
+const main = async () => {
+  // These components are exported as pure Svelte components.
+  const componentModules = {
+    'v-badge': await import('./elements/badge.svelte'),
```
Comment (mcous) [inline src/main.ts:24]: Can these become static imports since we removed the tag side effect?

**Review by mcous — APPROVED** [review]
Looks great! One question about the need for dynamic imports in main, but that can be moved to the restructuring PR

File: src/main.ts (hunk 7 above)
Comment (ethanlook) [reply src/main.ts:24]: Yup, that's totally reasonable. We'll do that when we restructure the package.
